src/compiler/extensions.ts

```diff
@@ -0,0 +1,330 @@
+         * @param {number} length The length of the error span
+         */
+        (shortname: string, err: string, start: number, length: number): void;
+    };
+    export type LintStopMethod = () => void;
+
+    /*
+    * Walkers call stop to halt recursion into the node's children
+    * Walkers call error to add errors to the output.
+    */
+    export interface LintWalker {
+        visit(node: Node, stop: LintStopMethod, error: LintErrorMethod): void;
+    }
+
+    export interface BaseProviderStatic {
+        readonly ["extension-kind"]: ExtensionKind;
+        new (state: {ts: typeof ts, args: any}): any;
+    }
+
+    export interface SyntacticLintProviderStatic extends BaseProviderStatic {
+        readonly ["extension-kind"]: ExtensionKind.SyntacticLint;
+        new (state: {ts: typeof ts, args: any, host: CompilerHost, program: Program}): LintWalker;
+    }
+
+    export interface SemanticLintProviderStatic extends BaseProviderStatic {
+        readonly ["extension-kind"]: ExtensionKind.SemanticLint;
+        new (state: {ts: typeof ts, args: any, host: CompilerHost, program: Program, checker: TypeChecker}): LintWalker;
+    }
+
+    export interface LanguageServiceHost {} // The members for these interfaces are provided in the services layer
+    export interface LanguageService {}
+    export interface LanguageServiceProvider {}
+    export interface DocumentRegistry {}
+
+    export interface LanguageServiceProviderStatic extends BaseProviderStatic {
+        readonly ["extension-kind"]: ExtensionKind.LanguageService;
+        new (state: { ts: typeof ts, args: any, host: LanguageServiceHost, service: LanguageService, registry: DocumentRegistry }): LanguageServiceProvider;
+    }
+
+    export namespace ExtensionKind {
+        export const SemanticLint: "semantic-lint" = "semantic-lint";
+        export type SemanticLint = "semantic-lint";
+        export const SyntacticLint: "syntactic-lint" = "syntactic-lint";
+        export type SyntacticLint = "syntactic-lint";
+        export const LanguageService: "language-service" = "language-service";
+        export type LanguageService = "language-service";
+    }
+    export type ExtensionKind = ExtensionKind.SemanticLint | ExtensionKind.SyntacticLint | ExtensionKind.LanguageService;
+
+    export interface ExtensionCollectionMap {
+        "syntactic-lint"?: SyntacticLintExtension[];
+        "semantic-lint"?: SemanticLintExtension[];
+        "language-service"?: LanguageServiceExtension[];
+        [index: string]: Extension[] | undefined;
+    }
+
+    export interface ExtensionBase {
+        name: string;
+        args: any;
+        kind: ExtensionKind;
+    }
+
+    export type Timestamp = number & { __timestampBrand: void };
+    export interface ProfileData {
+        globalBucket: string;
+        task: string;
+        start: Timestamp;
+        length?: Timestamp;
+    }
+
+    // @kind(ExtensionKind.SyntacticLint)

```

Comment:
It's an annotation we use throughout `types.ts` to indicate which enum member is associated with the `kind` field of a type... this may be the first time it's been used with string literal, though. It's probably unnecessary.